graph/src/util/stats.rs

```diff
@@ -0,0 +1,229 @@
+    }
+}
+
+/// Collect statistics over a moving window of size `window_size`. To keep
+/// the amount of memory needed to store the values inside the window
+/// constant, values are put into bins of size `bin_size`. For example, using
+/// a `window_size` of 5 minutes and a bin size of one second would use
+/// 300 bins. Each bin has constant size
+pub struct MovingStats {
+    window_size: Duration,
+    bin_size: Duration,
+    /// The buffer with measurements. The back has the most recent entries,
+    /// and the front has the oldest entries
+    bins: VecDeque<Bin>,
+    /// Sum over the values in `elements` The `start` of this bin
+    /// is meaningless
+    total: Bin,
+}
+
+/// Create `MovingStats` that use the window and bin sizes configured in
+/// the environment
+impl Default for MovingStats {
+    fn default() -> Self {
+        Self::new(*WINDOW_SIZE, *BIN_SIZE)
+    }
+}
+
+impl MovingStats {
+    pub fn new(window_size: Duration, bin_size: Duration) -> Self {
+        let capacity = if bin_size.as_millis() > 0 {
+            window_size.as_millis() as usize / bin_size.as_millis() as usize
+        } else {
+            1
+        };
+        MovingStats {
+            window_size,
+            bin_size,
+            bins: VecDeque::with_capacity(capacity),
+            total: Bin::new(Instant::now()),
+        }
+    }
+
+    /// Return `true` if the average of measurements in within `window_size`
+    /// is above `duration`
+    pub fn average_gt(&self, duration: Duration) -> bool {
+        // Depending on how often add() is called, we should
+        // call expire_bins first, but that would require taking a
+        // `&mut self`
+        self.total.average_gt(duration)
+    }
+
+    /// Return the average over the current window in milliseconds
+    pub fn average(&self) -> Option<Duration> {
+        self.total.duration.checked_div(self.total.count)
+    }
+
+    pub fn add(&mut self, duration: Duration) {
+        self.add_at(Instant::now(), duration);
+    }
+
+    /// Add an entry with the given timestamp. Note that the entry will
+    /// still be added either to the current latest bin or a new
+    /// latest bin. It is expected that subsequent calls to `add_at` still
+    /// happen with monotonically increasing `now` values. If the `now`
+    /// values do not monotonically increase, the average calculation
+    /// becomes imprecise because values are expired later than they
+    /// should be.
+    pub fn add_at(&mut self, now: Instant, duration: Duration) {
+        let need_new_bin = self
+            .bins
+            .back()
+            .map(|bin| now.saturating_duration_since(bin.start) >= self.bin_size)
+            .unwrap_or(true);
+        if need_new_bin {
+            self.bins.push_back(Bin::new(now));
+        }
+        self.expire_bins(now);
+        // unwrap is fine because we just added a bin if there wasn't one
+        // before
+        let bin = self.bins.back_mut().unwrap();
```

Comment:
Excellent point. I am addressing that by asserting in `MovingStats::new` that `0 < bin_size < window_size` since violating these constraints is very likely a configuration error.